mtec/integrationpluginmtec.cpp

```diff
@@ -0,0 +1,201 @@
+* along with this project. If not, see <https://www.gnu.org/licenses/>.
+*
+* For any further details and any questions please contact us under
+* contact@nymea.io or see our FAQ/Licensing Information on
+* https://nymea.io/license/faq
+*
+* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+
+#include "integrationpluginmtec.h"
+#include "plugininfo.h"
+
+IntegrationPluginMTec::IntegrationPluginMTec()
+{
+
+}
+
+void IntegrationPluginMTec::discoverThings(ThingDiscoveryInfo *info)
+{
+    qCDebug(dcMTec()) << "Discover M-Tec heat pumps";
+
+    if (info->thingClassId() == mtecThingClassId) {
+        QString description = "Heatpump";
+        ThingDescriptor descriptor(info->thingClassId(), "M-Tec", description);
+        info->addThingDescriptor(descriptor);
+
+        // TODO Find out, if a discovery is possible/needed
+        // Otherwise, just report no error for now
+        info->finish(Thing::ThingErrorNoError);
+    }
+}
+
+void IntegrationPluginMTec::setupThing(ThingSetupInfo *info)
+{
+    qCDebug(dcMTec()) << "Setup" << info->thing();
+
+    Thing *thing = info->thing();
+
+    if (thing->thingClassId() == mtecThingClassId) {
+        QHostAddress hostAddress = QHostAddress(thing->paramValue(mtecThingIpAddressParamTypeId).toString());
+
+        if (hostAddress.isNull()) {
+            info->finish(Thing::ThingErrorInvalidParameter, QT_TR_NOOP("No IP address given"));
+            return;
+        }
+
+        qCDebug(dcMTec()) << "User entered address: " << hostAddress.toString();
+
+        /* Check, if address is already in use for another device */
+        /* for (QHash<Thing *, MTec *>::iterator item=m_mtecConnections.begin(); item != m_mtecConnections.end(); item++) { */
+        /*     if (hostAddress.isEqual(item.value()->getHostAddress())) { */
+        /*         qCDebug(dcMTec()) << "Address of thing: " << item.value()->getHostAddress().toString(); */
+
+        /*         qCDebug(dcMTec()) << "Address in use already"; */
+        /*     } else { */
+        /*         qCDebug(dcMTec()) << "Different address of other thing: " << item.value()->getHostAddress().toString(); */
+
+        /*     } */
+        /* } */
+
+        foreach (MTec *mtecConnection, m_mtecConnections.values()) {
+            if (mtecConnection->getHostAddress().isEqual(hostAddress)) {
+                qCWarning(dcMTec()) << "Address" << hostAddress.toString() << "already in use by" << m_mtecConnections.key(mtecConnection);
+                info->finish(Thing::ThingErrorThingInUse, QT_TR_NOOP("IP address already in use by another thing."));
+                return;
+            }
+        }
+
+        qCDebug(dcMTec()) << "Creating M-Tec object";
+
+        /* Create new MTec object and store it in hash table */
+        MTec *mtec = new MTec(hostAddress, this);
+        m_mtecConnections.insert(thing, mtec);
+
+        QString msg = QT_TR_NOOP("Added M-Tec heatpump with address ");
```

Comment:
Unneeded string in translations, it will never be displayed, but will show up in translations.